### ISSUE TITLE
utils.py/sample_frames

### ISSUE THREAD

**Issue by FIRE-MOS** [taylover-pei/SSDG-CVPR2020#19]
single_video_frame_list[6 + j * frame_interval]) 
Does the above code mean we choose the fixed(固定的) one or two frames from each video,
and the rest frames of each video are abandoned?

Can I replace 6 with other number，such as 1,2,3?
Looking forward to your reply.

**Comment by Nam-pfiev1995** [comment]
Same question with you. I need to remove 6 to run. As mentioned in paper, they only used one frame by video. I think 6 is just a offset.

**Comment by FIRE-MOS** [comment]
since they only use one frame of video, I think there is no need to extract every frame of a video, so I extract one frame every ten frames, then I change 6 to 1, then the code can run successfully, the result is almost the same as the original paper result.

**Comment by Nam-pfiev1995** [comment]
Thanks for your suggestion. I also reproduce results near as in the paper by replacing 6 to 0.

**Comment by FIRE-MOS** [comment]
Did you test all four situations?
In I_C_M_to_O andO_C_I_to_M, the results are close to the paper result
but in O_C_M_to_I and O_M_I_to_C. the results are a little worse than  the paper result
The concrete result is as following:
![image](https://user-images.githubusercontent.com/63502295/99948377-55b1a200-2db4-11eb-9622-96d72a5407cc.png)

**Comment by FIRE-MOS** [comment]
while the original one is :
![image](https://user-images.githubusercontent.com/63502295/99948439-70841680-2db4-11eb-88f7-8ab438453c80.png)

**Comment by Nam-pfiev1995** [comment]
For me I test only O&C&M to I. My results are very close to the paper results.
![image](https://user-images.githubusercontent.com/72671625/99948951-8a086c80-2d7a-11eb-82d8-9ec334eb7b9d.png)

**Comment by FIRE-MOS** [comment]
Can you test the rest three situations?
I hope you can tell me all your results later.

**Comment by Nam-pfiev1995** [comment]
Sure, I will update all results when I finish running all experiments.

**Comment by FIRE-MOS** [comment]
Thanks in advance.

**Comment by roger-cv** [comment]
@FIRE-MOS 我遇到了和你相同的情况， O_C_M_to_I and O_M_I_to_C上效果差异比较大，其余两个测试和原论文接近。

**Comment by FIRE-MOS** [comment]
可以改超参数进行修改，效果可以逼近论文，差了一个百分点

**Comment by roger-cv** [comment]
@FIRE-MOS 改哪部分的超参数呢？能请教一下吗？另外，能看一下你现在的训练结果吗？

**Comment by FIRE-MOS** [comment]
你比较一下四个condition的config.py就知道了
 改动之后的O_C_M_to_I： 
![image](https://user-images.githubusercontent.com/63502295/100184931-911bb000-2f1d-11eb-9d25-da07030c4ba5.png)
超参不好调，没什么规律，硬调出来的

**Comment by roger-cv** [comment]
额，好的，谢谢。

**Comment by hufangjian** [comment]
你好，请问一下这个测试自己的数据集怎么测试？

**Comment by FIRE-MOS** [comment]
@roger-cv 你跑一个完整的大约需要多久啊，我用的是一个1080ti，需要2到3个小时
@hufangjian 我不知道啊，应该是从dg_test中改吧

**Comment by roger-cv** [comment]
@FIRE-MOS 我用v100，需要6个小时，这好像是假的v100，哈哈哈

**Comment by FIRE-MOS** [comment]
你是说4个experiments的总时间吗，我说的2到3个是单个的，比如I&C&MtoO

**Comment by roger-cv** [comment]
我是一次性跑4个experiments哦

**Comment by FIRE-MOS** [comment]
那还差不多，你重新调整了一下代码吗，我只能一次跑一个

**Comment by roger-cv** [comment]
不是，我的意思是我跑四个实验的总耗时差不多六个小时。

**Comment by FIRE-MOS** [comment]
懂了，也是手动一个一个跑的

**Comment by roger-cv** [comment]
是的呢

**Comment by FIRE-MOS** [comment]
fine！20年的central difference convolution那篇你跑了吗

**Comment by roger-cv** [comment]
cdc效果非常差，我不知道paper中的效果为什么那么好

**Comment by FIRE-MOS** [comment]
ssdg这篇文章训练和测试都是在一个视频里面取一或两帧，这样是不是太少了啊

**Comment by roger-cv** [comment]
不是吧，一个数据集差不多要有5w张图片吧

**Comment by FIRE-MOS** [comment]
每个视频取一帧啊，代码就是这样设计的，论文中也讲了
![image](https://user-images.githubusercontent.com/63502295/100496180-12c54500-318d-11eb-9736-cbdb7658d4ff.png)
所以每个数据集的图片数就是视频的数目，比如casia就只有几百张，而oulu就有几千张

**Comment by roger-cv** [comment]
我说的是我总的数据集的大小，但是代码中会再次对数据集进行采样的。

**Comment by roger-cv** [comment]
训练的时候还是和原论文一样的

**Comment by FIRE-MOS** [comment]
你是提前每个视频都把每一帧都保存，还是每隔几帧保存

**Comment by roger-cv** [comment]
都有，不同数据集处理不一样的

**Comment by FIRE-MOS** [comment]
大的隔多帧，小的数据集隔少帧

**Comment by roger-cv** [comment]
记不得了，我四个数据集最后生成差不多20w张图片，我处理数据也是花了很长时间

**Comment by FIRE-MOS** [comment]
我是10w张

**Comment by roger-cv** [comment]
OK

**Comment by taylover-pei** [comment]
single_video_frame_list[6 + j * frame_interval])这里面的6当时仅仅是为了剔除前面的几帧视频而已，可以随意修改。

你们讨论的挺热闹，有问题的话，再提一个issue吧。

**Comment by XinyingZheng** [comment]
> @FIRE-MOS 我遇到了和你相同的情况， O_C_M_to_I and O_M_I_to_C上效果差异比较大，其余两个测试和原论文接近。

想问一下，在函数Real_AdLoss中，def Real_AdLoss(discriminator_out, criterion, shape_list):
    # generate ad_label
    ad_label1_index = torch.LongTensor(shape_list[0], 1).fill_(0)
    ad_label1 = ad_label1_index.cuda()
    ad_label2_index = torch.LongTensor(shape_list[1], 1).fill_(1)
    ad_label2 = ad_label2_index.cuda()
    ad_label3_index = torch.LongTensor(shape_list[2], 1).fill_(2)
    ad_label3 = ad_label3_index.cuda()
    ad_label = torch.cat([ad_label1, ad_label2, ad_label3], dim=0).view(-1)

**Comment by XinyingZheng** [comment]
> @FIRE-MOS 我遇到了和你相同的情况， O_C_M_to_I and O_M_I_to_C上效果差异比较大，其余两个测试和原论文接近。

想问一下，在函数 Real_AdLoss中，def Real_AdLoss(discriminator_out, criterion, shape_list):
    # generate ad_label
    ad_label1_index = torch.LongTensor(shape_list[0], 1).fill_(0)
    ad_label1 = ad_label1_index.cuda()
    ad_label2_index = torch.LongTensor(shape_list[1], 1).fill_(1)
    ad_label2 = ad_label2_index.cuda()
    ad_label3_index = torch.LongTensor(shape_list[2], 1).fill_(2)
    ad_label3 = ad_label3_index.cuda()
    ad_label = torch.cat([ad_label1, ad_label2, ad_label3], dim=0).view(-1)        真人的对抗标签为啥是0，1，2，那假体的label是多少呢，我现在有自己的数据，想用这个单边域泛化的思想优化模型

**Comment by chin-sheng** [comment]
假体不會丟到函数 Real_AdLoss中 只會針對真人樣本作優化
